File: LeetCodeExample.Test/Tree/_01466_ReorderRoutesToMakeAllPathsLeadToCityZero.cs
using Microsoft.VisualStudio.TestPlatform.ObjectModel;
using NUnit.Framework;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Runtime.InteropServices;
namespace LeetCodeExample.Test;

/// <summary>
//There are n cities numbered from 0 to n - 1 and n - 1 roads such that there is only one way to travel between two different cities(this network form a tree). Last year, The ministry of transport decided to orient the roads in one direction because they are too narrow.
//Roads are represented by connections where connections[i] = [ai, bi] represents a road from city ai to city bi.

//This year, there will be a big event in the capital (city 0), and many people want to travel to this city.

//Your task consists of reorienting some roads such that each city can visit the city 0. Return the minimum number of edges changed.

//It's guaranteed that each city can reach city 0 after reorder.
/// </summary>
public class _01466_ReorderRoutesToMakeAllPathsLeadToCityZero
{
    Dictionary<int, List<(int, int)>> adjList = new();
    bool[] visited;

    public int MinReorder(int n, int[][] connections)
    {
        visited = new bool[n];
        for (int i = 0; i < n; i++)
        {
            adjList.Add(i, new List<(int, int)>());
        }
        for (int i = 0; i < connections.Length; i++)
        {
            adjList[connections[i][0]].Add((connections[i][1], 1));
            adjList[connections[i][1]].Add((connections[i][0], 0));
        }

        visited[0] = true;
        return dfs(0);
    }

    int dfs(int i)
    {
        int count = 0;
        // Roads going away need to be flipped, add one to each of these
        foreach ((int n, int cost) in adjList[i])
        {
            if (visited[n])
                continue;
            visited[n] = true;
            count+= cost;
            count += dfs(n);
        }
        return count;
    }
}

File: LeetCodeExample.Test/Graph/_01466_ReorderRoutesToMakeAllPathsLeadToCityZero.py
# There are n cities numbered from 0 to n - 1 and n - 1 roads such that there is only one way to travel between two different cities (this network form a tree). Last year, The ministry of transport decided to orient the roads in one direction because they are too narrow.

# Roads are represented by connections where connections[i] = [ai, bi] represents a road from city ai to city bi.

# This year, there will be a big event in the capital (city 0), and many people want to travel to this city.

# Your task consists of reorienting some roads such that each city can visit the city 0. Return the minimum number of edges changed.

# It's guaranteed that each city can reach city 0 after reorder.'


# This isn't my answer
from typing import List
class Solution:
    def minReorder(self, n: int, connections: List[List[int]]) -> int:
        adjList = {}
        for i in range(n):
            adjList[i] = []
        for connection in connections:
            # Roads going away have a cost
            # Make it bidirectional so we can just start at 0 and reverse roads
            # as we pass
            adjList[connection[0]].append((connection[1], 1))
            adjList[connection[1]].append((connection[0], 0))
        visited = [False] * n

        def dfs(city: int) -> int:
            count = 0
            for otherCity, cost in adjList[city]:
                if visited[otherCity]:
                    continue
                visited[otherCity] = True
                count += cost
                count += dfs(otherCity)
            return count

        visited[0] = True
        return dfs(0)